### PR TITLE
Governance Upgrade New

### PR DESCRIPTION
**Pull request by yubo-ruan**
### New Updates:

- [x] search `// OLD:` for original code
- [x] add test cases on VoteToken and GovernorAlpha
- [x] all test passed
- [x] fix warnings 
- [x] fix license issue
- [x] rebase to dev branch and move contracts to a new folders
- [x]  remove @title, @author
- [x] add proxy behind governor and timelock contracts

2. Hal and I decided to keep the pragma experimental ABIEncoderV2 

### VoteToken

- [x] use solidity 0.6.10 version
- [x] override _transfer function (add _moveDelegates)
- [x] fix bugs with keccak256(abi.encodePacked())
- [x] abstract contract VoteToken is TimeLockedToken 
- [x] uint96(balanceOf[delegator]) explicit conversion
- [x] unlockedBalance(delegator)

### ProxyStorage

**Add additional storage of** 

- [x] mapping delegates
- [x] struct Checkpoint
- [x] mapping checkpoints
- [x] mapping numCheckpoints
- [x] mapping nonces

### Timelock.sol

- [x] using solidity ^0.6.10
- [x] fix receive() external payable { }
- [ ] add admin_initialized

### TrustToken

- [x] Inherit from VoteToken

### GovernorAlpha.sol

- [x] name change to  "TrustToken Governor Alpha"
- [x] using solidity ^0.6.10
- [x] add MIT License
- [x] quorumVotes changes to 58000000e8      (4% of Tru)
- [x] proposalThreshold changes to 14500000e8      (1% of Tru)
- [x] fix TrustTokenInterface and trustToken instance

### Notice:
1. GovernorAlpha.sol, Timelock.sol, and part of VoteToken are forked from [Compound](https://github.com/compound-finance/compound-protocol/blob/master/contracts/Governance/Comp.sol)

### CODE REVIEW

File: test/governance/GovernorAlpha.test.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        expect(await governorAlpha.latestProposalIds(initialHolder.address)).to.eq(1)
+        await governorAlpha.connect(owner).cancel(1) //gudian can cancel a proposal
+        expect(await governorAlpha.state(1)).to.eq(2)
+      })
+    })
+  })
+  describe('castVote', () => {
+    beforeEach(async() => {
+      await governorAlpha.connect(initialHolder).propose(target,values,signatures,callDatas,description)
+      await provider.send('evm_mine', []) //mine one block
```
Comment (hal909) [inline test/governance/GovernorAlpha.test.ts:96]: replace evm_mine with time travel

File: contracts/governance/GovernorAlpha.sol [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    function proposalThreshold() public pure returns (uint) { return 14500000e8; } // 14,500,000 = 1% of TRU
+
+    // @notice The maximum number of actions that can be included in a proposal
+    function proposalMaxOperations() public pure returns (uint) { return 10; } // 10 actions
+
+    // @notice The delay before voting on a proposal may take place, once proposed
+    function votingDelay() public pure returns (uint) { return 1; } // 1 block
+
+    // @notice The duration of voting on a proposal, in blocks
+    function votingPeriod() public pure returns (uint) { return 17280; } // ~3 days in blocks (assuming 15s blocks)
```
Comment (hal909) [inline contracts/governance/GovernorAlpha.sol:36]: create ability for timelock to change the voting period. Store votingPeriod as a variable

**Review by hal909 — CHANGES_REQUESTED** [review]
- Put GovernanceAlpha and Timelock contracts behind an OwnedUpgradeabilityProxy
- Make governance own both contracts in tests (only governance can make upgrades_
- Add more tests
- Add a test for upgrading proxies
- Add a test for submitting proposals (make sure the values are correct in submissions)

**Review by yubo-ruan — COMMENTED** [review]
It seems like the execution will revert when the target is the governorAlpha contract itself. I think we can use proxy to update to a new governor contract as an alternative

File: contracts/governance/GovernorAlpha.sol (hunk 2 above)
Comment (yubo-ruan) [reply contracts/governance/GovernorAlpha.sol:36]: It seems like the execution will revert when the target is the governorAlpha contract itself. I think we can use proxy to update to a new governor contract as an alternative

**Review by hal909 — APPROVED** [review]
looks great! I can help add more tests, but let's move onto the veTRU contract tomorrow

**Comment by yubo-ruan** [comment]
Thanks! I’ll chat with you tomorrow

Best,
Yubo Ruan
________________________________

**Review by vanruch — APPROVED** [review]

File: contracts/governance/GovernorAlpha.sol [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    // @notice An event emitted when a proposal has been queued in the Timelock
+    event ProposalQueued(uint id, uint eta);
+
+    // @notice An event emitted when a proposal has been executed in the Timelock
+    event ProposalExecuted(uint id);
+
+    /**
+     * @dev Initialize sets the addresses of timelock contract, trusttoken contract, and guardian
+     */
+    function initialize(address timelock_, address trustToken_, address guardian_, uint _votingPeriod) external {
```
Comment (barrutko) [inline contracts/governance/GovernorAlpha.sol:152]: Any reason why do we have `timelock_`, `trustToken_`, `guardian_`, but `_votingPeriod`?
(`_` placing)

**Pull request merged**